horace_core/sqw/@sqw/set_efix.m

```diff
@@ -23,53 +23,79 @@
 
 % Perform operations
 % ==================
-narg=numel(varargin);
-if narg<1 || narg>2
+if ~(isnumeric(efix) && numel(efix)>=1 && all(isfinite(efix)) && all(efix>=0))
     error('HORACE:sqw:invalid_argument',...
-        'This function accepts one or two input arguments')
+        'efix must be numeric scalar or array of finite values');
 end
-if narg>=1
-    efix=varargin{1}(:);
-    if ~(isnumeric(efix) && numel(efix)>=1 && all(isfinite(efix)) && all(efix>=0))
-        error('HORACE:sqw:invalid_argument',...
-            'efix must be numeric scalar or array of finite values');
-    end
-end
-if narg>=2
-    emode=varargin{2};
+if ~exist('emode','var')
+    emode=[];   % indicates emode to be left untouched
+else
     if ~(isnumeric(emode) && isscalar(emode) && (emode==0||emode==1||emode==2))
         error('HORACE:sqw:invalid_argument',...
             'emode must 1 (direct geometry), 2 (indirect geometry) or 0 (elastic)')
     end
-else
-    emode=[];   % indicates emode to be left untouched
 end
 
 
 % Change efix and emode
 % ---------------------
 
 % Check the number of spe files matches the number of efix
+% check what kind of efix array is provided and beeing set.
+% single for all objects, change each object or change all
+% objects
 nefix=numel(efix);
-if nefix>1
-    for i=1:numel(obj)
-        if obj(i).experiment_info.n_runs ~=nefix
+n_runs_in_obj = arrayfun(@(x)x.experiment_info.n_runs,obj);
+eq_nefix = n_runs_in_obj == nefix;
+split_emode = false;
+if all(eq_nefix)
+    split_runs = false;
+elseif nefix == sum(n_runs_in_obj) && nefix ~=1
+    split_runs = true;
+    if ~isempty(emode)
+        if numel(emode) ~= nefix
             error('HORACE:sqw:invalid_argument',...
-                ['An array of efix values was given but its length (%d) ',...
-                'does not match the number of spe files (%d) in the sqw N:%d source being altered'],...
-                nefix,obj(i).experiment_info.n_runs,i)
+                'Array of efix and emodes are provided, but the length of efix array (%d) is different from the length of emode array (%d)',...
+                nefix,numel(emode))
         end
+        split_emode = true;
+    else % numel(emode) == 1;
+        split_emode = false;
     end
+elseif nefix == 1
+    split_runs = false;
+else
+    error('HORACE:sqw:invalid_argument',...
+        ['An array of efix values was given but its length (%d) ',...
+        'does not match the number of runs either in every object (%d) or in all sqw objects together (%d)'],...
+        nefix,n_runs_in_obj(1),sum(n_runs_in_obj));
 end
```

Comment:
```suggestion
if ~all(eq_nefix) && (nefix ~= 1 || nefix ~= sum(n_runs_in_obj))
    error('HORACE:sqw:invalid_argument',...
          ['An array of efix values was given but its length (%d) ',...
           'does not match the number of runs either in every object (%d) or in all sqw objects together (%d)'],...
          nefix,n_runs_in_obj(1),sum(n_runs_in_obj));
end

if ~isempty(emode) && numel(emode) ~= nefix
    error('HORACE:sqw:invalid_argument',...
          'Array of efix and emodes are provided, but the length of efix array (%d) is different from the length of emode array (%d)',...
          nefix,numel(emode))
end

split_runs = ~all(eq_nefix) && nefix == sum(n_runs_in_obj) && nefix ~= 1;
split_emode = split_runs && ~isempty(emode);
```

Reply:
modified for clarity from different point of view. The previous code was confusing and I am not sure the substitution works fine.